ee/katc/deserialize_firefox.go

```diff
@@ -436,6 +442,222 @@ func deserializeArray(arrayLength uint32, srcReader *bytes.Reader) ([]byte, erro
+	lengthFromByteLength, err := byteLengthToTypedArrayLength(arrayType, byteLength)
+	if err != nil {
+		return nil, fmt.Errorf("converting byte length to array length: %w", err)
+	}
+	if length == math.MaxUint64 {
+		length = lengthFromByteLength
+	}
+
+	// Read in the TypedArray: byteLength raw data bytes, then byte offset, then finally the padding.
+	rawTypedArrayBytes := make([]byte, byteLength)
+	for i := 0; i < int(byteLength); i++ {
+		b, err := srcReader.ReadByte()
+		if err != nil {
+			return nil, fmt.Errorf("reading byte %d of %d in TypedArray: %w", i, byteLength, err)
+		}
+		rawTypedArrayBytes[i] = b
+	}
+	var byteOffset uint64
+	if err := binary.Read(srcReader, binary.NativeEndian, &byteOffset); err != nil {
+		return nil, fmt.Errorf("reading byteoffset in TypedArray: %w", err)
+	}
+
+	// Handle padding -- data is packed into uint64_t words. See StructuredClone's ComputePadding.
+	// We read and discard this data.
+	elemSize, err := elementSize(arrayType)
+	if err != nil {
+		return nil, fmt.Errorf("getting element size to calculate padding: %w", err)
+	}
+	leftoverLength := (lengthFromByteLength % 8) * elemSize
+	padding := -leftoverLength & 7
+	for i := 0; i < int(padding); i++ {
+		if _, err := srcReader.ReadByte(); err != nil {
+			return nil, fmt.Errorf("reading byte %d of %d of padding after TypedArray: %w", i, padding, err)
+		}
+	}
+
+	// We have all the data we need from srcReader to handle the TypedArray.
+	// Create a new reader for our raw TypedArray bytes.
+	typedArrayReader := bytes.NewReader(rawTypedArrayBytes)
+
+	// If we have an offset at the start of the TypedArray, read and discard that number of bytes.
+	for i := 0; i < int(byteOffset); i++ {
+		if _, err := typedArrayReader.ReadByte(); err != nil {
+			return nil, fmt.Errorf("reading byte %d of %d of offset in TypedArray: %w", i, byteOffset, err)
+		}
+	}
+
+	// Finally! We have the raw data that we can interpret as the correct type for the TypedArray,
+	// we've handled padding, we've handled the offset, and we know how many elements are in our array.
+	// We can proceed with reinterpreting the data in typedArrayReader as the correct TypedArray.
+	var results any
+	switch arrayType {
+	case typedArrayInt8:
+		int8Arr := make([]int8, length)
+		for i := 0; i < int(length); i++ {
+			if err := binary.Read(typedArrayReader, binary.NativeEndian, &int8Arr[i]); err != nil {
+				return nil, fmt.Errorf("reading %T at index %d in TypedArray: %w", int8Arr[i], i, err)
+			}
+		}
+		results = int8Arr
+	case typedArrayUint8, typedArrayUint8Clamped:
+		uint8Arr := make([]uint8, length)
+		for i := 0; i < int(length); i++ {
+			if err := binary.Read(typedArrayReader, binary.NativeEndian, &uint8Arr[i]); err != nil {
+				return nil, fmt.Errorf("reading %T at index %d in TypedArray: %w", uint8Arr[i], i, err)
+			}
+		}
+		results = uint8Arr
+	case typedArrayInt16:
+		int16Arr := make([]int16, length)
+		for i := 0; i < int(length); i++ {
```

Comment:
just a q because i think I'm misunderstanding something- why don't we need to manipulate the length for these to get the expected number of iterations based on the types being read in, similar to e.g. `readFloat64Array`?

Reply:
ahh okay i see, that is what i missed. thank you!